### DISCUSSION TITLE
Multiple helics_broker at one computer for parallel operation

### DISCUSSION THREAD

**Discussion by asinghal91** [GMLC-TDC/HELICS#1574] [Q&A] [answered]
@kdheepak 
In a machine learning application, we need to instantiate multiple helics broker simultaneously to run in parallel in 1 computer. I am running helics on windows. Just to test if a manual brokerport works or not, I tried following changes in a working example of co-simulation with 1 broker and 2 federates (python and gridlabd). Note that I run my example using a run.bat (attached). I am using helics 2.4.0.

- I added --type=zmq and --port=20200 in my original helics_broker line:
`helics_broker --type=zmq --port=20200 --federates 2 --loglevel 3`
 
- updated config.json of both federates (python and gridlabd) by adding a brokerport variable:
```
{
  "name": "GLD",
  "loglevel": 3,
  "coreType": "zmq",
  "brokerport": "20200",
  "period": 1.0,
  "wait_for_current_time_update": true,
  ...
}
 ```
I get following error from gridlabd:
ERROR    [INIT] : helics_msg::init: an unexpected error occurred when trying to create a CombinationFederate using the configuration string: 
gld_helics_confg.json
This is the error that was caught:
Value is not convertible to Int64.
ERROR    [INIT] : init_by_deferral(): object GLD initialization failed
ERROR    [INIT] : model initialization failed
FATAL    [INIT] : shutdown after simulation stopped prematurely
FATAL    [INIT] : environment startup failed: Invalid argument

[config_file.txt](https://github.com/GMLC-TDC/HELICS/files/5222330/config_file.txt)
[gld_helics_confg.txt](https://github.com/GMLC-TDC/HELICS/files/5222332/gld_helics_confg.txt)
[run.txt](https://github.com/GMLC-TDC/HELICS/files/5222342/run.txt)

**Comment by phlptp** [comment]
The first step would be to look at 
https://docs.helics.org/en/latest/user-guide/simultaneous_cosimulations.html

If you are doing this a bunch of times all on the same machine I would suggest the broker_server It was designed for this use case, so you don't have to restart a single broker over and over again and coordinate all the ports. 

As for this particular issue
try changing
```
"brokerport": "20200",
```
to 
```
"brokerport": 20200,
```

This particular field has generated enough errors we should probably support a string in that field but it reads as an int currently.

**Reply by asinghal91** [reply]
when using `"brokerport": 20200,` I am getting following error:

```
ERROR    [INIT] : helics_msg::init: an unexpected error occurred when trying to create a CombinationFederate using the configuration string: 
gld_helics_confg.json
This is the error that was caught:
Unable to connect to broker->unable to register federate
ERROR    [INIT] : init_by_deferral(): object GLD initialization failed
ERROR    [INIT] : model initialization failed
FATAL    [INIT] : shutdown after simulation stopped prematurely
FATAL    [INIT] : environment startup failed: Invalid argument
```
I get exact same error when I try broker_server using `helics_broker_server --type=zmq --federates 2 --loglevel 3` and removing `brokerport: 20200` from both federate config files.

Is it possible that the error is because I am running it on Windows ? Is there any working example available ?

**Comment by asinghal91** [comment, marked as answer]
@phlptp : I finally got it work. Here is the fix:

- the line argument for the broker is actually `--brokerport=20200` 

- and variable in the federate config file is `“brokerPort”: 20200 `

However, I am not sure how to use helics_broker_server ? Can it be instantiated using helicsRunner or python interface or some other method that allow me to instantiate helics_brokers from a python file ?
